File: duckdata/crosslink/python/shared_memory/cpp_wrapper.py
"""
C++ binding wrapper for CrossLink Python

This module loads the C++ bindings for CrossLink and provides an interface
to the core C++ implementation for better cross-language zero-copy performance.
"""
import os
import sys
import importlib.util
from pathlib import Path
import warnings
# Add pyarrow import for type hinting if needed
try:
    import pyarrow as pa
except ImportError:
    pa = None

# Define variables for C++ binding availability
_cpp_available = False
_cpp_module = None

# Try to import the C++ extension module
def _try_import_cpp():
    """Try to import the C++ module from various locations"""
    global _cpp_available, _cpp_module
    
    # If already loaded, don't try again
    if _cpp_available:
        return True
    
    try:
        # First method: try direct import if installed in Python path
        import crosslink.crosslink_cpp as cpp_mod
        _cpp_module = cpp_mod
        _cpp_available = True
        return True
    except ImportError:
        pass
    
    # Second method: try to load from local path
    try:
        # Get the potential paths
        current_dir = Path(__file__).parent.resolve()
        project_dir = current_dir.parent.parent
        
        # Find the platform-specific extension suffix (e.g., .so, .pyd)
        try:
            import sysconfig
            so_suffix = sysconfig.get_config_var('EXT_SUFFIX') # e.g., .cpython-310-darwin.so
            if not so_suffix:
                so_suffix = '.so' # Fallback for Linux/macOS generic
        except:
            so_suffix = '.so' # Default fallback
        pyd_suffix = '.pyd' # Windows specific

        possible_paths = [
            # Check with platform specific suffix first
            current_dir / f"crosslink_cpp{so_suffix}",
            current_dir / f"crosslink_cpp{pyd_suffix}",
            project_dir / "cpp" / "build" / f"crosslink_cpp{so_suffix}",
            project_dir / "cpp" / "build" / f"crosslink_cpp{pyd_suffix}",
            # Fallback to generic names
            current_dir / "crosslink_cpp.so",  # Linux/macOS
            current_dir / "crosslink_cpp.pyd",  # Windows
            project_dir / "cpp" / "build" / "crosslink_cpp.so",
            project_dir / "cpp" / "build" / "crosslink_cpp.pyd",
            project_dir / "cpp" / "crosslink_cpp.so",
            project_dir / "cpp" / "crosslink_cpp.pyd"
        ]
        
        # --- Debugging path checking --- (Removed)
        # print("--- [cpp_wrapper.py] Debug: Checking paths ---")
        # for p_debug in possible_paths:
        #     print(f"  Checking: {p_debug} | Exists: {p_debug.exists()}")
        # print("--- [cpp_wrapper.py] Debug: End checking paths ---")
        # # --- End Debugging ---

        for path in possible_paths:
            if path.exists():
                # Load using importlib
                try:
                    # print(f"--- [cpp_wrapper.py] Debug: Attempting to load {path} ---") # Removed
                    spec = importlib.util.spec_from_file_location("crosslink_cpp", path)
                    if spec is None:
                        # print(f"--- [cpp_wrapper.py] Debug: spec_from_file_location returned None for {path} ---") # Removed
                        continue # Try next path

                    if spec.loader is None:
                        # print(f"--- [cpp_wrapper.py] Debug: spec.loader is None for {path} ---") # Removed
                        continue # Try next path
                        
                    cpp_mod = importlib.util.module_from_spec(spec)
                    if cpp_mod is None:
                        # print(f"--- [cpp_wrapper.py] Debug: module_from_spec returned None for {path} ---") # Removed
                        continue # Try next path
                        
                    spec.loader.exec_module(cpp_mod)
                    # print(f"--- [cpp_wrapper.py] Debug: Successfully loaded {path} ---") # Removed
                    _cpp_module = cpp_mod
                    _cpp_available = True
                    return True
                except Exception as e_load:
                    # Print the detailed loading error if needed, but keep it minimal for production
                    warnings.warn(f"Warning: Failed during importlib loading of '{path}'. Error: {e_load}") 
                    continue # Try next path
    
    # Handle outer exceptions during path setup if necessary
    except Exception as e:
        warnings.warn(f"Warning: Outer exception in _try_import_cpp. Error: {e}")
        pass # Let it fall through to the final warning if all else fails
    
    # If we get here, no method succeeded
    warnings.warn("C++ bindings not found in any expected location, falling back to Python implementation")
    return False

# Try to import when module is loaded
_try_import_cpp()

class CrossLinkCpp:
    """Python wrapper for the C++ CrossLink implementation"""
    
    def __init__(self, db_path="crosslink.duckdb", debug=False):
        """Initialize the C++ CrossLink instance
        
        Args:
            db_path: Path to the database file
            debug: Enable debug logging
        
        Raises:
            ImportError: If C++ bindings are not available
        """
        global _cpp_available, _cpp_module
        
        if not _cpp_available:
            if not _try_import_cpp():  # Try one more time
                raise ImportError("CrossLink C++ bindings are not available")
        
        # Ensure PyArrow types are available if using C++ bindings
        if pa is None:
             warnings.warn("PyArrow is not installed. C++ bindings require PyArrow.")
             raise ImportError("PyArrow is required for CrossLink C++ bindings.")
        
        self._cpp_impl = _cpp_module.CrossLinkCpp(db_path, debug)
        self.db_path = db_path
        self.debug = debug
        # Expose the bound StreamWriter and StreamReader classes directly if needed
        # Or rely on the return types of push/pull_stream
        self.StreamWriter = getattr(_cpp_module, "StreamWriter", None)
        self.StreamReader = getattr(_cpp_module, "StreamReader", None)
    
    def push(self, table: 'pa.Table', name: str = "", description: str = "") -> str:
        """Push a PyArrow table to CrossLink
        
        Args:
            table: PyArrow Table
            name: Optional name for the dataset
            description: Optional description
        
        Returns:
            str: Dataset ID
        
        Raises:
            RuntimeError: If push operation fails
        """
        try:
            return self._cpp_impl.push(table, name, description)
        except Exception as e:
            if self.debug:
                print(f"C++ push operation failed: {e}")
            raise RuntimeError(f"Failed to push data: {e}")
    
    def pull(self, identifier: str) -> 'pa.Table':
        """Pull a dataset as a PyArrow table
        
        Args:
            identifier: Dataset ID or name
        
        Returns:
            pyarrow.Table: Arrow table
        
        Raises:
            RuntimeError: If pull operation fails
        """
        try:
            return self._cpp_impl.pull(identifier)
        except Exception as e:
            if self.debug:
                print(f"C++ pull operation failed: {e}")
            raise RuntimeError(f"Failed to pull data for {identifier}: {e}")
    
    def query(self, sql: str) -> 'pa.Table':
        """Execute a SQL query and return results as a PyArrow table
        
        Args:
            sql: SQL query string
        
        Returns:
            pyarrow.Table: Query results
            
        Raises:
            RuntimeError: If query operation fails
        """
        try:
            return self._cpp_impl.query(sql)
        except Exception as e:
            if self.debug:
                print(f"C++ query operation failed: {e}")
            raise RuntimeError(f"Failed to execute query: {e}")
    
    def list_datasets(self) -> list[str]:
        """List available datasets
        
        Returns:
            list: List of dataset IDs
            
        Raises:
            RuntimeError: If list operation fails
        """
        try:
            return self._cpp_impl.list_datasets()
        except Exception as e:
            if self.debug:
                print(f"C++ list_datasets operation failed: {e}")
            raise RuntimeError(f"Failed to list datasets: {e}")
    
    def register_notification(self, callback) -> str:
        """Register a callback for dataset notifications
        
        Args:
            callback: Function to call when a dataset changes
        
        Returns:
            str: Registration ID
            
        Raises:
            RuntimeError: If registration fails
        """
        try:
            return self._cpp_impl.register_notification(callback)
        except Exception as e:
            if self.debug:
                print(f"C++ register_notification operation failed: {e}")
            raise RuntimeError(f"Failed to register notification: {e}")
    
    def unregister_notification(self, registration_id: str):
        """Unregister a notification callback
        
        Args:
            registration_id: Registration ID from register_notification
            
        Raises:
            RuntimeError: If unregistration fails
        """
        try:
            return self._cpp_impl.unregister_notification(registration_id)
        except Exception as e:
            if self.debug:
                print(f"C++ unregister_notification operation failed: {e}")
            raise RuntimeError(f"Failed to unregister notification: {e}")
    
    def cleanup(self):
        """Clean up all resources
        
        Raises:
            RuntimeError: If cleanup fails
        """
        try:
            return self._cpp_impl.cleanup()
        except Exception as e:
            if self.debug:
                print(f"C++ cleanup operation failed: {e}")
            raise RuntimeError(f"Failed to cleanup resources: {e}")
    
    def get_from_shared_memory(self, shared_memory_key, metadata=None):
        """Get data from shared memory
        
        Args:
            shared_memory_key: Shared memory key
            metadata: Optional metadata about the shared memory
            
        Returns:
            pyarrow.Table or None: Retrieved data or None if not found
            
        Raises:
            RuntimeError: If operation fails (but not if key not found)
        """
        try:
            # Handle None or empty key
            if shared_memory_key is None or shared_memory_key == "":
                if self.debug:
                    print("Cannot get from shared memory: key is None or empty")
                return None
                
            return self._cpp_impl.get_from_shared_memory(shared_memory_key, metadata or {})
        except KeyError:
            # Specific exception for key not found
            if self.debug:
                print(f"Shared memory key not found: {shared_memory_key}")
            return None
        except Exception as e:
            if self.debug:
                print(f"C++ get_from_shared_memory operation failed: {e}")
            raise RuntimeError(f"Failed to get data from shared memory: {e}")
    
    def setup_shared_memory(self, table, dataset_id):
        """Set up shared memory for a table
        
        Args:
            table: PyArrow table
            dataset_id: ID to associate with the shared memory
        
        Returns:
            str: Shared memory key
        
        Raises:
            RuntimeError: If setup fails
        """
        try:
            return self._cpp_impl.setup_shared_memory(table, dataset_id)
        except Exception as e:
            if self.debug:
                print(f"C++ setup_shared_memory operation failed: {e}")
            raise RuntimeError(f"Failed to setup shared memory: {e}")
    
    def cleanup_shared_memory(self, shared_memory_key=None):
        """Clean up shared memory regions
        
        Args:
            shared_memory_key: Specific key to clean up, or None for all
            
        Returns:
            int: Number of regions cleaned up
            
        Raises:
            RuntimeError: If cleanup fails
        """
        try:
            return self._cpp_impl.cleanup_shared_memory(shared_memory_key)
        except Exception as e:
            if self.debug:
                print(f"C++ cleanup_shared_memory operation failed: {e}")
            raise RuntimeError(f"Failed to cleanup shared memory: {e}")

    # === New Streaming Methods ===

    def push_stream(self, schema: 'pa.Schema', name: str = "") -> tuple[str, object]:
        """Start pushing a stream of Arrow RecordBatches.

        Args:
            schema: The pyarrow.Schema for the data that will be streamed.
            name: Optional name for the stream. A unique ID will be generated if empty.

        Returns:
            tuple[str, StreamWriter]: A tuple containing the stream ID and the
                                      StreamWriter object to write batches to.

        Raises:
            RuntimeError: If starting the stream fails.
            ImportError: If C++ bindings or required StreamWriter class not found.
        """
        if not self._cpp_impl or not self.StreamWriter:
             raise ImportError("Streaming requires available C++ bindings and StreamWriter class.")
        try:
            # The C++ binding returns std::pair<string, shared_ptr<StreamWriter>>
            # pybind11 converts this to a Python tuple(str, StreamWriter_bound_object)
            stream_id, writer_obj = self._cpp_impl.push_stream(schema, name)
            return stream_id, writer_obj
        except Exception as e:
            if self.debug:
                print(f"C++ push_stream operation failed: {e}")
            raise RuntimeError(f"Failed to start stream: {e}")

    def pull_stream(self, stream_id: str) -> object:
        """Connect to an existing stream to pull Arrow RecordBatches.

        Args:
            stream_id: The ID of the stream to connect to.

        Returns:
            StreamReader: A StreamReader object to read batches from.
                          The reader is iterable (for batch in reader: ...).

        Raises:
            RuntimeError: If connecting to the stream fails (e.g., not found).
            ImportError: If C++ bindings or required StreamReader class not found.
        """
        if not self._cpp_impl or not self.StreamReader:
             raise ImportError("Streaming requires available C++ bindings and StreamReader class.")
        try:
             # The C++ binding returns shared_ptr<StreamReader>
             # pybind11 converts this to a Python StreamReader_bound_object
            reader_obj = self._cpp_impl.pull_stream(stream_id)
            return reader_obj
        except Exception as e:
            if self.debug:
                print(f"C++ pull_stream operation failed: {e}")
            # Distinguish between 'not found' and other errors if C++ provides specific exceptions
            # For now, raise a general error.
            raise RuntimeError(f"Failed to connect to stream '{stream_id}': {e}")

# Expose the check function at the module level
def is_cpp_available():
    """Check if the C++ bindings for CrossLink are loaded and available"""
    global _cpp_available
    return _cpp_available 

File: duckdata/crosslink/julia/shared_memory/cpp_wrapper.jl
"""
C++ bindings wrapper for CrossLink in Julia

This module loads and provides an interface to the C++ implementation of CrossLink
for better cross-language interoperability and performance.
"""
module CrossLinkCppWrapper

using Arrow
using DuckDB
using DataFrames
export CrossLinkCpp, is_cpp_available
export StreamWriter, StreamReader, push_stream, pull_stream

# Flag for C++ binding availability
global _cpp_available = false
global _cpp_lib = nothing

"""
    try_load_cpp()

Try to load the C++ library from various possible locations.
Returns true if successful, false otherwise.
"""
function try_load_cpp()
    global _cpp_available, _cpp_lib
    
    # Return early if already loaded
    if _cpp_available
        return true
    end
    
    try
        # Get potential paths for the library
        project_dir = joinpath(dirname(@__DIR__), "..")
        
        # Different platforms have different library extensions
        lib_name = Sys.iswindows() ? "libcrosslink_jl.dll" : 
                  Sys.isapple() ? "libcrosslink_jl.dylib" : "libcrosslink_jl.so"
        
        # Look in multiple possible locations
        possible_paths = [
            joinpath(@__DIR__, lib_name),
            joinpath(project_dir, "julia", "deps", lib_name),
            joinpath(project_dir, "cpp", "build", lib_name),
            joinpath(project_dir, "deps", lib_name)
        ]
        
        # Try to load the library from each path
        for path in possible_paths
            if isfile(path)
                # Use CxxWrap if available
                try
                    # Check if CxxWrap is available
                    using CxxWrap
                    
                    # Load the library
                    @info "Loading C++ library from $path"
                    _cpp_lib = CxxWrap.load_library(path)
                    
                    # Set flag to true
                    _cpp_available = true
                    return true
                catch e
                    # Try loading with ccall if CxxWrap fails
                    try
                        # Direct ccall loading as fallback
                        _cpp_lib = path
                        
                        # Try a test function to verify the library works
                        # For example, get the version number
                        test_ptr = dlsym(dlopen(_cpp_lib), :crosslink_get_version)
                        if test_ptr != C_NULL
                            _cpp_available = true
                            return true
                        end
                    catch inner_e
                        @warn "Failed to load C++ library with ccall: $inner_e"
                    end
                end
            end
        end
        
        # If we get here, we failed to load the library
        return false
    catch e
        @warn "Error trying to load C++ bindings: $e"
        return false
    end
end

"""
    is_cpp_available()

Check if C++ bindings are available.
"""
function is_cpp_available()
    global _cpp_available
    return _cpp_available
end

"""
    has_shared_memory()

Check if the C++ implementation has shared memory support.
"""
function has_shared_memory()
    if !is_cpp_available()
        return false
    end
    
    # In a real implementation, we would check if shared memory
    # is supported in the C++ implementation
    return true
end

"""
    init_cpp()

Initialize the C++ wrapper.
"""
function init_cpp()
    global _cpp_available
    
    # Try to load C++ bindings
    _cpp_available = try_load_cpp()
    
    return _cpp_available
end

# Initialize on module load
init_cpp()

"""
    CrossLinkCpp

Julia wrapper for the C++ CrossLink implementation.
"""
mutable struct CrossLinkCpp
    db_path::String
    debug::Bool
    cpp_instance::Any  # Either a CxxWrap object or ccall-based handle
    
    """
        CrossLinkCpp(db_path::String="crosslink.duckdb", debug::Bool=false)
        
    Create a new CrossLink instance using C++ bindings.
    """
    function CrossLinkCpp(db_path::String="crosslink.duckdb", debug::Bool=false)
        # Try to load C++ bindings if not already loaded
        if !try_load_cpp()
            error("C++ bindings are not available")
        end
        
        global _cpp_lib
        
        # Create the C++ object based on what loading method worked
        if typeof(_cpp_lib) <: String  # ccall path
            # Create C++ object via ccall
            instance_ptr = ccall(
                dlsym(dlopen(_cpp_lib), :crosslink_create), 
                Ptr{Cvoid}, 
                (Cstring, Bool), 
                db_path, 
                debug
            )
            
            # Create the Julia wrapper
            instance = new(db_path, debug, instance_ptr)
            
            # Set finalizer to clean up C++ resources
            finalizer(x -> ccall(
                dlsym(dlopen(_cpp_lib), :crosslink_destroy),
                Cvoid,
                (Ptr{Cvoid},),
                x.cpp_instance
            ), instance)
            
            return instance
        else  # CxxWrap path
            # Create C++ object via CxxWrap
            cpp_obj = _cpp_lib.CrossLinkCpp(db_path, debug)
            
            # Create the Julia wrapper
            instance = new(db_path, debug, cpp_obj)
            
            return instance
        end
    end
end

"""
    push(cl::CrossLinkCpp, table::Arrow.Table, name::String="", description::String="")

Push an Arrow table to CrossLink via C++ bindings.
"""
function push(cl::CrossLinkCpp, table::Arrow.Table, name::String="", description::String="")
    global _cpp_lib
    
    # Use the appropriate method based on how the library was loaded
    if typeof(cl.cpp_instance) <: Ptr{Cvoid}  # ccall path
        # Serialize Arrow table to bytes
        io = IOBuffer()
        Arrow.write(io, table)
        arrow_bytes = take!(io)
        
        # Call C++ function
        result_ptr = ccall(
            dlsym(dlopen(_cpp_lib), :crosslink_push),
            Cstring,
            (Ptr{Cvoid}, Ptr{UInt8}, Csize_t, Cstring, Cstring),
            cl.cpp_instance,
            arrow_bytes,
            length(arrow_bytes),
            name,
            description
        )
        
        # Convert result to Julia string
        result = unsafe_string(result_ptr)
        
        # Free the C string
        ccall(
            dlsym(dlopen(_cpp_lib), :crosslink_free_string),
            Cvoid,
            (Cstring,),
            result_ptr
        )
        
        return result
    else  # CxxWrap path
        # Convert Arrow table to format expected by binding
        io = IOBuffer()
        Arrow.write(io, table)
        arrow_bytes = take!(io)
        
        # Call the C++ method
        return cl.cpp_instance.push(arrow_bytes, name, description)
    end
end

"""
    pull(cl::CrossLinkCpp, identifier::String)

Pull a dataset from CrossLink as an Arrow table via C++ bindings.
"""
function pull(cl::CrossLinkCpp, identifier::String)
    global _cpp_lib
    
    # Use the appropriate method based on how the library was loaded
    if typeof(cl.cpp_instance) <: Ptr{Cvoid}  # ccall path
        # Call C++ function to get Arrow data
        result_ptr = ccall(
            dlsym(dlopen(_cpp_lib), :crosslink_pull),
            Ptr{UInt8},
            (Ptr{Cvoid}, Cstring, Ptr{Csize_t}),
            cl.cpp_instance,
            identifier,
            Ref{Csize_t}(0)  # Will be filled with data size
        )
        
        # Get the size
        size_ptr = Ref{Csize_t}(0)
        ccall(
            dlsym(dlopen(_cpp_lib), :crosslink_get_last_data_size),
            Cvoid,
            (Ptr{Cvoid}, Ptr{Csize_t}),
            cl.cpp_instance,
            size_ptr
        )
        
        # Copy and convert data to Arrow table
        data_size = size_ptr[]
        data = unsafe_wrap(Array, result_ptr, data_size)
        io = IOBuffer(copy(data))
        
        # Free the C++ memory
        ccall(
            dlsym(dlopen(_cpp_lib), :crosslink_free_data),
            Cvoid,
            (Ptr{UInt8},),
            result_ptr
        )
        
        # Read arrow table from memory
        return Arrow.Table(io)
    else  # CxxWrap path
        # Call the C++ method
        result_bytes = cl.cpp_instance.pull(identifier)
        
        # Convert bytes to Arrow table
        io = IOBuffer(result_bytes)
        return Arrow.Table(io)
    end
end

"""
    query(cl::CrossLinkCpp, sql::String)

Execute SQL query via C++ bindings and return results as Arrow table.
"""
function query(cl::CrossLinkCpp, sql::String)
    global _cpp_lib
    
    # Similar implementation as pull but with query method
    if typeof(cl.cpp_instance) <: Ptr{Cvoid}  # ccall path
        # Call C++ function to get Arrow data
        result_ptr = ccall(
            dlsym(dlopen(_cpp_lib), :crosslink_query),
            Ptr{UInt8},
            (Ptr{Cvoid}, Cstring, Ptr{Csize_t}),
            cl.cpp_instance,
            sql,
            Ref{Csize_t}(0)  # Will be filled with data size
        )
        
        # Get the size
        size_ptr = Ref{Csize_t}(0)
        ccall(
            dlsym(dlopen(_cpp_lib), :crosslink_get_last_data_size),
            Cvoid,
            (Ptr{Cvoid}, Ptr{Csize_t}),
            cl.cpp_instance,
            size_ptr
        )
        
        # Copy and convert data to Arrow table
        data_size = size_ptr[]
        data = unsafe_wrap(Array, result_ptr, data_size)
        io = IOBuffer(copy(data))
        
        # Free the C++ memory
        ccall(
            dlsym(dlopen(_cpp_lib), :crosslink_free_data),
            Cvoid,
            (Ptr{UInt8},),
            result_ptr
        )
        
        # Read arrow table from memory
        return Arrow.Table(io)
    else  # CxxWrap path
        # Call the C++ method
        result_bytes = cl.cpp_instance.query(sql)
        
        # Convert bytes to Arrow table
        io = IOBuffer(result_bytes)
        return Arrow.Table(io)
    end
end

"""
    list_datasets(cl::CrossLinkCpp)

List available datasets via C++ bindings.
"""
function list_datasets(cl::CrossLinkCpp)
    global _cpp_lib
    
    if typeof(cl.cpp_instance) <: Ptr{Cvoid}  # ccall path
        # Call C++ function
        result_ptr = ccall(
            dlsym(dlopen(_cpp_lib), :crosslink_list_datasets),
            Ptr{Ptr{UInt8}},
            (Ptr{Cvoid}, Ptr{Csize_t}),
            cl.cpp_instance,
            Ref{Csize_t}(0)  # Will be filled with count
        )
        
        # Get the count
        count_ptr = Ref{Csize_t}(0)
        ccall(
            dlsym(dlopen(_cpp_lib), :crosslink_get_list_count),
            Cvoid,
            (Ptr{Cvoid}, Ptr{Csize_t}),
            cl.cpp_instance,
            count_ptr
        )
        
        count = count_ptr[]
        
        # Extract each string
        datasets = String[]
        for i in 1:count
            str_ptr = unsafe_load(result_ptr, i)
            push!(datasets, unsafe_string(str_ptr))
        end
        
        # Free the C++ memory
        ccall(
            dlsym(dlopen(_cpp_lib), :crosslink_free_string_array),
            Cvoid,
            (Ptr{Ptr{UInt8}}, Csize_t),
            result_ptr,
            count
        )
        
        return datasets
    else  # CxxWrap path
        # Call the C++ method
        return cl.cpp_instance.list_datasets()
    end
end

"""
    register_notification(cl::CrossLinkCpp, callback::Function)

Register a callback for dataset notifications via C++ bindings.
"""
function register_notification(cl::CrossLinkCpp, callback::Function)
    global _cpp_lib
    
    if typeof(cl.cpp_instance) <: Ptr{Cvoid}  # ccall path
        # Not easily implemented with ccall, would need custom C wrapper
        error("Notification callbacks not supported with ccall interface")
    else  # CxxWrap path
        # Call the C++ method
        return cl.cpp_instance.register_notification(callback)
    end
end

"""
    unregister_notification(cl::CrossLinkCpp, registration_id::String)

Unregister a notification callback via C++ bindings.
"""
function unregister_notification(cl::CrossLinkCpp, registration_id::String)
    global _cpp_lib
    
    if typeof(cl.cpp_instance) <: Ptr{Cvoid}  # ccall path
        ccall(
            dlsym(dlopen(_cpp_lib), :crosslink_unregister_notification),
            Cvoid,
            (Ptr{Cvoid}, Cstring),
            cl.cpp_instance,
            registration_id
        )
    else  # CxxWrap path
        # Call the C++ method
        cl.cpp_instance.unregister_notification(registration_id)
    end
    
    return nothing
end

"""
    cleanup(cl::CrossLinkCpp)

Clean up resources via C++ bindings.
"""
function cleanup(cl::CrossLinkCpp)
    global _cpp_lib
    
    if typeof(cl.cpp_instance) <: Ptr{Cvoid}  # ccall path
        ccall(
            dlsym(dlopen(_cpp_lib), :crosslink_cleanup),
            Cvoid,
            (Ptr{Cvoid},),
            cl.cpp_instance
        )
    else  # CxxWrap path
        # Call the C++ method
        cl.cpp_instance.cleanup()
    end
    
    return nothing
end

# StreamWriter class to handle writing batches to a stream
mutable struct StreamWriter
    cpp_writer::Any
    schema::Arrow.Schema
    stream_id::String
    
    function StreamWriter(cpp_writer, schema, stream_id)
        instance = new(cpp_writer, schema, stream_id)
        
        # Set finalizer to clean up C++ resources if needed
        if has_shared_memory()
            finalizer(x -> try_close(x), instance)
        end
        
        return instance
    end
end

# StreamReader class to handle reading batches from a stream
mutable struct StreamReader
    cpp_reader::Any
    schema::Union{Arrow.Schema, Nothing}
    stream_id::String
    
    function StreamReader(cpp_reader, schema, stream_id)
        instance = new(cpp_reader, schema, stream_id)
        
        # Set finalizer to clean up C++ resources if needed
        if has_shared_memory()
            finalizer(x -> try_close(x), instance)
        end
        
        return instance
    end
end

# Helper functions for StreamWriter
function write_batch(writer::StreamWriter, batch::Arrow.Table)
    if typeof(writer.cpp_writer) <: Ptr{Cvoid}  # ccall path
        error("Not yet implemented for direct ccall")
    else  # CxxWrap path
        # Serialize the batch to send to C++
        io = IOBuffer()
        Arrow.write(io, batch)
        batch_bytes = take!(io)
        
        # Call the C++ method
        writer.cpp_writer.write_batch(batch_bytes)
        return nothing
    end
end

function close(writer::StreamWriter)
    if typeof(writer.cpp_writer) <: Ptr{Cvoid}  # ccall path
        error("Not yet implemented for direct ccall")
    else  # CxxWrap path
        writer.cpp_writer.close()
        return nothing
    end
end

function try_close(writer::StreamWriter)
    try
        close(writer)
    catch e
        # Ignore errors during finalization
    end
end

function get_stream_id(writer::StreamWriter)
    return writer.stream_id
end

function get_schema(writer::StreamWriter)
    return writer.schema
end

# Helper functions for StreamReader
function read_next_batch(reader::StreamReader)
    if typeof(reader.cpp_reader) <: Ptr{Cvoid}  # ccall path
        error("Not yet implemented for direct ccall")
    else  # CxxWrap path
        # Get the batch bytes from C++
        batch_bytes = reader.cpp_reader.read_next_batch()
        
        # If we got an empty array, there are no more batches
        if length(batch_bytes) == 0
            return nothing
        end
        
        # Deserialize the batch
        io = IOBuffer(batch_bytes)
        return Arrow.Table(io)
    end
end

function close(reader::StreamReader)
    if typeof(reader.cpp_reader) <: Ptr{Cvoid}  # ccall path
        error("Not yet implemented for direct ccall")
    else  # CxxWrap path
        reader.cpp_reader.close()
        return nothing
    end
end

function try_close(reader::StreamReader)
    try
        close(reader)
    catch e
        # Ignore errors during finalization
    end
end

function get_stream_id(reader::StreamReader)
    return reader.stream_id
end

function get_schema(reader::StreamReader)
    if reader.schema === nothing
        if typeof(reader.cpp_reader) <: Ptr{Cvoid}  # ccall path
            error("Not yet implemented for direct ccall")
        else  # CxxWrap path
            # Get the schema bytes from C++
            schema_bytes = reader.cpp_reader.schema()
            
            # Deserialize the schema
            io = IOBuffer(schema_bytes)
            reader.schema = Arrow.Table(io).schema
        end
    end
    
    return reader.schema
end

# Define iterator interface for StreamReader
function Base.iterate(reader::StreamReader, state=nothing)
    batch = read_next_batch(reader)
    if batch === nothing
        return nothing
    else
        return (batch, nothing)
    end
end

function Base.IteratorSize(::Type{StreamReader})
    return Base.SizeUnknown()
end

function Base.eltype(::Type{StreamReader})
    return Arrow.Table
end

"""
    push_stream(cl::CrossLinkCpp, schema::Arrow.Schema, name::String="")

Start a new stream with the given schema using C++ bindings.
Returns a tuple of (stream_id, StreamWriter).
"""
function push_stream(cl::CrossLinkCpp, schema::Arrow.Schema, name::String="")
    if !has_shared_memory()
        error("C++ shared memory implementation is not available")
    end
    
    global _cpp_lib
    
    # Use the appropriate method based on how the library was loaded
    if typeof(cl.cpp_instance) <: Ptr{Cvoid}  # ccall path
        error("Not yet implemented for direct ccall")
    else  # CxxWrap path
        # Serialize the schema to send to C++
        io = IOBuffer()
        Arrow.write(io, Arrow.Table(; a=Int64[])) # Empty table with schema
        schema_bytes = take!(io)
        
        # Call the C++ method
        result = cl.cpp_instance.push_stream(schema_bytes, name)
        
        # Extract stream_id and writer
        stream_id = result[1]
        cpp_writer = result[2]
        
        # Create Julia StreamWriter wrapper
        writer = StreamWriter(cpp_writer, schema, stream_id)
        
        return (stream_id, writer)
    end
end

"""
    pull_stream(cl::CrossLinkCpp, stream_id::String)

Connect to an existing stream with the given ID using C++ bindings.
Returns a StreamReader.
"""
function pull_stream(cl::CrossLinkCpp, stream_id::String)
    if !has_shared_memory()
        error("C++ shared memory implementation is not available")
    end
    
    global _cpp_lib
    
    # Use the appropriate method based on how the library was loaded
    if typeof(cl.cpp_instance) <: Ptr{Cvoid}  # ccall path
        error("Not yet implemented for direct ccall")
    else  # CxxWrap path
        # Call the C++ method
        cpp_reader = cl.cpp_instance.pull_stream(stream_id)
        
        # Create Julia StreamReader wrapper
        # We'll get the schema later when needed
        reader = StreamReader(cpp_reader, nothing, stream_id)
        
        return reader
    end
end

# Try to load C++ bindings when module is loaded
try_load_cpp()

end # module CrossLinkCppWrapper 